python-threatexchange/threatexchange/fetcher/fetch_state_base.py

```diff
@@ -0,0 +1,123 @@
+#!/usr/bin/env python3
+# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
+
+"""
+Base classes for passing data between SignalExchangeAPIs and other interfaces.
+
+Many implementations will choose to extend these to add additional metadata
+needed to power API features.
+"""
+
+
+from dataclasses import dataclass
+from enum import Enum
+import typing as t
+
+from threatexchange.signal_type.signal_base import SignalType
+
+
+TFetchStateCheckpoint = t.TypeVar("TFetchStateCheckpoint")  # TODO for now
+
+
+class SignalOpinionCategory(Enum):
+    """
+    What the opinion on a signal is.
+
+    Some APIs may not support all of these, but each of these should influence
+    what action you might take as a result of matching, otherwise it might
+    make more sense as a tag
+    """
+
+    FALSE_POSITIVE = 0  # Signal generates false positives
+    WORTH_INVESTIGATING = 1  # Indirect indicator
+    TRUE_POSITIVE = 2  # Confirmed meets category
+
+
+@dataclass
+class SignalOpinion:
+    """
+    The metadata of a single signal upload.
+
+    Certain APIs won't have any concept of owner, category, or tags,
+    in which case owner=0, category=TRUE_POSITIVE, tags=[] is reasonable
+    default.
+    """
+
+    owner: int
+    category: SignalOpinionCategory
+    tags: t.List[str]
+
+
+@dataclass
+class FetchedSignalData:
+    """
+    Metadata to make decisions on matches and power feedback on the fetch API.
+    """
+
+    collab_name: str  # Corresponds to the collab config that fetched it
+    opinions: t.List[SignalOpinion]
+
+
+class FetchDelta:
+    """
+    Contains the result of a fetch.
+
+    Needs only to be interpretable by FetchedState
+    """
+
+    def record_count(self) -> int:
+        """Helper for --limit"""
+        return 1
+
+    def next_checkpoint(self) -> TFetchStateCheckpoint:
+        """
+        A serializable checkpoint for fetch
+        """
+        raise NotImplementedError
+
+
+class FetchedState:
+    """
+    An interface to previously fetched or persisted state.
+    """
+
+    def get_checkpoint(self) -> TFetchStateCheckpoint:
+        """
+        Returns the last checkpoint passed to merge() after a flush()
+        """
+        raise NotImplementedError
+
+    def merge(self, subsequent: FetchDelta) -> None:
+        """
+        Merge a subsequent FetchDelta into this one.
+
+        At the implementation's discretion, it may call flush() or the
+        equivalent work.
+        """
+        raise NotImplementedError
+
+    def flush(self) -> None:
+        """
```

Comment:
nit/ignore/question: does returning None here have a meaning and why does the base class do so instead of throwing?

Reply:
Derp, you are right, this should be NotImplemented, since there's expectations on the interface that are not satisfied by a default implementation of `None`